### DISCUSSION TITLE
Unexpected result with proxy routes.

### DISCUSSION THREAD

**Discussion by lostb1t** [salvo-rs/salvo#363] [General]
When I proxy a route like this

```
            .push(
                Router::with_path("/library/metadata/<id>")
                    .hoop(disable_related_query)
                    .handle(proxy.clone()),
            )
```
The request url of `http://localhost/library/metadata/500` will result in an proxied request of `http://upstream`, without an path. While you would expect the path to be included in the proxy request: aka `http://upstream/library/metadata/500`

**Comment by chrislearn** [comment]
only the part `<**rest>` in router path will added to upstream path.

For example:

```rust
 .push(
                Router::with_path("/library/metadata/<**rest>")
                    .hoop(disable_related_query)
                    .handle(proxy.clone()),
            )
```

when you request url is `http://localhost/library/metadata/500`, it will proxy to `http://upstream/500`.

**Comment by lostb1t** [comment]
Yeah I find that unexpected. I would expect the whole path to be proxied. Not only the rest part

**Comment by lostb1t** [comment]
What's the suggested for doing this then? 

I need to add different middleware to different proxy routes. I don't want to apply it to every routes

```
        router = router
            .push(
                Router::new().path("/library/metadata/<id>/related")
                    .hoop(Timeout::new(Duration::from_secs(5)))
                    .handle(proxy.clone()),
            )
            .push(
                Router::with_path("/library/metadata/<id>")
                    .hoop(disable_related_query)
                    .handle(proxy.clone()),
            )
            .push(
                Router::with_path("/playQueues")
                    .hoop(another_middleware)
                    .handle(proxy.clone()),
            );
 ```

**Comment by chrislearn** [comment]
You can use a middlware to put `<**rest>` to request's params. But this way is not elegant.

**Comment by lostb1t** [comment]
yeah that's the only solution for now. Not really dry tho, as I have to apply my own path filters. But at least I know that this is by design.

**Comment by chrislearn** [comment]
Maybe I can add a function to allow set custom way to build the target url. You can check the proxy branch. 


```rust
pub fn default_url_rest_getter(req: &Request, _depot: &Depot) -> String {
    let param = req.params().iter().find(|(key, _)| key.starts_with('*'));
    let mut rest = if let Some((_, rest)) = param {
        encode_url_path(rest)
    } else {
        "".into()
    };
    if let Some(query) = req.uri().query() {
        rest = format!("{}?{}", rest, query);
    }
    rest
}

    /// Set url rest getter.
    #[inline]
    pub fn url_rest_getter<G>(mut self, url_rest_getter: G) -> Self
    where
        G: Fn(&Request, &Depot) -> String + Send + Sync + 'static,
    {
        self.url_rest_getter = Box::new(url_rest_getter);
        self
    }

```

**Comment by lostb1t** [comment]
yeah that works. Im already overloading exactly that part. But this is much cleaner

**Comment by lostb1t** [comment]
Give me a ping when this is merged so I can switch the branch

**Comment by chrislearn** [comment]
@sarendsen This is merged, Version 0.50.3 released.
